### PR TITLE
Add wait_for_all_acked support

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Related to ros2/rmw#295

### CODE REVIEW

File: rclcpp/test/rclcpp/test_publisher.cpp [hunk 1]
```diff
@@ -536,3 +537,45 @@ TEST_F(TestPublisher, get_network_flow_endpoints_errors) {
     EXPECT_NO_THROW(publisher->get_network_flow_endpoints());
   }
 }
+
+TEST_F(TestPublisher, check_wait_for_all_acked_return) {
```
Comment (fujitatomoya) [inline rclcpp/test/rclcpp/test_publisher.cpp:541]: i see this is unit test for interfaces in rclcpp, but i think the following test cases are `nice to have` to make sure all the call stack works fine.

- actually issue `rcl_publisher_wait_for_all_acked` on publisher w/o `mocking_utils::patch_and_return`.
- QoS setting for BestEffort and Reliable since those behaves differently in implementation.

File: rclcpp/include/rclcpp/publisher_base.hpp [hunk 2]
```diff
@@ -203,6 +204,69 @@ class PublisherBase : public std::enable_shared_from_this<PublisherBase>
...
+   * If the timeout is 0 then this function will be non-blocking; checking all published message
+   * data were acknowledged (If acknowledged, return true. Otherwise, return false),
+   * but not waiting.
+   * If the timeout is greater than 0 then this function will return after that period of time has
+   * elapsed (return false) or all published message data were acknowledged (return true).
+   *
+   * \param[in] timeout the duration to wait for all published message data were acknowledged.
+   * \return `true` if all published message data were acknowledged before timeout, otherwise
+   *   `false`.
+   * \throws rclcpp::exceptions::RCLError if middleware doesn't support or internal error occurs
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp:224]: Yeah. I will update description on exception

File: rclcpp/test/rclcpp/test_publisher.cpp (hunk 1 above)
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_publisher.cpp:541]: Yes, I should add test to confirm call stack works well.

**Comment by Barry-Xu-2018** [comment]
Friendly ping @fujitatomoya.  Please check again.

**Comment by fujitatomoya** [comment]
CC: @ivanpauno @wjwwood friendly ping.

**Comment by fujitatomoya** [comment]
@ros-pull-request-builder retest this please

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15207/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9927/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12887/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15389/)

File: rclcpp/include/rclcpp/publisher_base.hpp [hunk 3]
```diff
@@ -203,6 +204,73 @@ class PublisherBase : public std::enable_shared_from_this<PublisherBase>
...
+    // https://stackoverflow.com/a/44637334/2089061
+    // However, this doesn't solve the issue for all possible duration types of period.
+    // Follow-up issue: https://github.com/ros2/rclcpp/issues/1177
+    constexpr auto ns_max_as_double =
+      std::chrono::duration_cast<std::chrono::duration<double, std::chrono::nanoseconds::period>>(
+      maximum_safe_cast_ns);
+    if (timeout > ns_max_as_double) {
+      throw std::invalid_argument{
+              "timeout must be less than std::chrono::nanoseconds::max()"};
+    }
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/publisher_base.hpp:254]: according to std::chrono::duration_cast() documentation in cppreference:

> Casting from a floating-point duration to an integer duration is subject to undefined behavior when the floating-point value is NaN, infinity, or too large to be representable by the target's integer type. Otherwise, casting to an integer duration is subject to truncation as with any static_cast to an integer type.

so I see the point of this, but the code it's a bit hard to follow and we need that in other places.
I would rather have a helper function in `rcpputils` "safe_duration_cast()`, and use that here.

For the moment, I would rather use std::chrono::duration_cast(), as we're already doing the same elsewhere.

Comment (ivanpauno) [reply rclcpp/include/rclcpp/publisher_base.hpp:254]: I also think the code can be simplified:

```suggestion
    if (timeout > std::chrono::nanoseconds::max()) {
      throw std::invalid_argument{
              "timeout must be less than std::chrono::nanoseconds::max()"};
    }
```

see https://en.cppreference.com/w/cpp/chrono/duration/operator_cmp.

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 i will try another CI after all @ivanpauno 's comments are addressed.

**Comment by Barry-Xu-2018** [comment]
> @Barry-Xu-2018 i will try another CI after all @ivanpauno 's comments are addressed.

Okay. I will address @ivanpauno 's comments.

File: rclcpp/include/rclcpp/publisher_base.hpp (hunk 3 above)
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp:254]: I think simplified codes may introduce some bugs.  
I will further confirm it.

Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp:254]: If timeout is  std::chrono::hours(10000000), it can pass (timeout > std::chrono::nanoseconds::max()).   
But we will get  a negative value after converting as below code  
int64_t v = (std::chrono::duration_cast<std::chrono::nanoseconds>(timeout)).count()  

So if use simplified way, it will include bugs.  
 I tend to adopt your below suggestion.  What do you think ? @ivanpauno 

> I would rather have a helper function in rcpputils "safe_duration_cast()`, and use that here.

Comment (ivanpauno) [reply rclcpp/include/rclcpp/publisher_base.hpp:254]: > If timeout is std::chrono::hours(10000000), it can pass (timeout > std::chrono::nanoseconds::max()).
But we will get a negative value after converting as below code

How?
`timeout > std::chrono::nanoseconds::max()` is true so we will throw an exception and the cast will never happen.

There's a problem if `timeout > std::chrono::nanoseconds::max()` is false and the cast fails anyway (because of rounding issues). I'm really not sure of that, cppreference doesn't document any guarantee.

> I tend to adopt your below suggestion. What do you think ? @ivanpauno

SGTM

Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp:254]: >> If timeout is std::chrono::hours(10000000), it can pass (timeout > std::chrono::nanoseconds::max()).
>> But we will get a negative value after converting as below code
>
> How?
> timeout > std::chrono::nanoseconds::max() is true so we will throw an exception and the cast will never happen.

Sorry. Make you confused. 
I means `timeout > std::chrono::nanoseconds::max()` is false. We get  a negative value after the cast is called.

Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp:254]: >> I tend to adopt your below suggestion. What do you think ? @ivanpauno
>
> SGTM

I created a PR ros2/rcpputils#145. Please help to review.

**Comment by Barry-Xu-2018** [comment]
Friendly ping @fujitatomoya

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15248/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9967/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12930/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15442/)

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya 

Sorry. I want you to confirm the modification of ros2/rcpputils#145.  
I wrong to add comment to this issue.

**Comment by Barry-Xu-2018** [comment]
Wait for new interface `rcpputils::convert_to_nanoseconds` merged ros2/rcpputils#145

**Comment by Barry-Xu-2018** [comment]
Do rebase.    

ros2/rcpputils#145 (merged) isn't included in CI system. So build failed.

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 some comments from @ivanpauno still needs to be addressed. once they are all fixed, i will retry CI.

File: rclcpp/include/rclcpp/publisher_base.hpp [hunk 4]
```diff
@@ -203,6 +204,73 @@ class PublisherBase : public std::enable_shared_from_this<PublisherBase>
   std::vector<rclcpp::NetworkFlowEndpoint>
   get_network_flow_endpoints() const;
 
+  /// Wait until all published message data is acknowledged or until the specified timeout elapses.
+  /**
+   * This function waits until all published message data were acknowledged by all subscribers or
+   * timeout.
+   *
+   * timeout must be less than std::chrono::nanoseconds::max().
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp:213]: Yes.
I will remove it. The limitation was described at `\throws std::invalid_argument`.

**Comment by Barry-Xu-2018** [comment]
> @Barry-Xu-2018 some comments from @ivanpauno still needs to be addressed. once they are all fixed, i will retry CI.

Thanks for your kind reminder. I missed some review comments.

**Comment by Barry-Xu-2018** [comment]
Do rebase.

**Review by ivanpauno — APPROVED** [review]
I have some nitpicking in the docs.
LGTM with green CI and comments addressed!

**Comment by Barry-Xu-2018** [comment]
@ivanpauno 

Thanks for your comments.  
I have addressed them.

**Comment by Barry-Xu-2018** [comment]
@ros-pull-request-builder retest this please

File: rclcpp/include/rclcpp/publisher_base.hpp [hunk 5]
```diff
@@ -203,6 +205,46 @@ class PublisherBase : public std::enable_shared_from_this<PublisherBase>
   std::vector<rclcpp::NetworkFlowEndpoint>
   get_network_flow_endpoints() const;
 
+  /// Wait until all published messages are acknowledged or until the specified timeout elapses.
+  /**
+   * This method waits until all published messages are acknowledged by all matching
+   * subscriptions or the given timeout elapses.
+   *
+   * If the timeout is negative then this method will block indefinitely until all published
+   * message data are acknowledged.
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/publisher_base.hpp:214]:
```suggestion
   * messages are acknowledged.
```

I think you can either say "messages are" or "message data is".
My preference is the first option.
I might be wrong, as I'm not a native speaker.

File: rclcpp/include/rclcpp/publisher_base.hpp [hunk 6]
```diff
@@ -203,6 +205,46 @@ class PublisherBase : public std::enable_shared_from_this<PublisherBase>
...
+   * If the timeout is zero then this method will not block, it will check if all published
+   * messages were acknowledged and return immediately.
+   * If the timeout is greater than zero, this method will wait until all published messages are
+   * acknowledged or the timeout elapses.
+   *
+   * This method only waits for acknowledgments if the publisher's QoS profile is RELIABLE.
+   * Otherwise this method will immediately return `true`.
+   *
+   * \param[in] timeout the duration to wait for all published message data to be acknowledged.
+   * \return `true` if all published message data were acknowledged before the given timeout
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/publisher_base.hpp:224]:
```suggestion
   * \param[in] timeout the duration to wait for all published messages to be acknowledged.
   * \return `true` if all published messages were acknowledged before the given timeout
```

Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp:224]: Okay.  
I will check if the contents has the similar  problems and resolve this type of problem once and for all.

File: rclcpp/include/rclcpp/publisher_base.hpp (hunk 5 above)
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/publisher_base.hpp]: Agree.  
I'm not a native speaker too :)

**Comment by Barry-Xu-2018** [comment]
Friendly ping @ivanpauno

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15810/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10488/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13472/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16091/)

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 uncrustify failures are not related to your fix, but just in case could you rebase? after that i will start CI again, and expecting that it turns into all green.

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya 

Rebase was done.   

The failure of uncrustify as below. They should be fixed by another PR.
```
5: Code style divergence in file 'include/rclcpp/wait_set_policies/sequential_synchronization.hpp':
5: 
5: --- include/rclcpp/wait_set_policies/sequential_synchronization.hpp
5: +++ include/rclcpp/wait_set_policies/sequential_synchronization.hpp.uncrustify
5: @@ -206 +206 @@
5: -      void(std::shared_ptr<rclcpp::Waitable>&&, std::shared_ptr<void> &&)
5: +      void(std::shared_ptr<rclcpp::Waitable>&&, std::shared_ptr<void>&&)
5: 
5: Code style divergence in file 'include/rclcpp/wait_set_policies/thread_safe_synchronization.hpp':
5: 
5: --- include/rclcpp/wait_set_policies/thread_safe_synchronization.hpp
5: +++ include/rclcpp/wait_set_policies/thread_safe_synchronization.hpp.uncrustify
5: @@ -219 +219 @@
5: -      void(std::shared_ptr<rclcpp::Waitable>&&, std::shared_ptr<void> &&)
5: +      void(std::shared_ptr<rclcpp::Waitable>&&, std::shared_ptr<void>&&)
5: 
5: Code style divergence in file 'test/utils/rclcpp_gtest_macros.hpp':
5: 
5: --- test/utils/rclcpp_gtest_macros.hpp
5: +++ test/utils/rclcpp_gtest_macros.hpp.uncrustify
5: @@ -165 +165 @@
5: -      is_the_result_of_the_throwing_expression_equal_to_the_expected_throwable = \
5: +    is_the_result_of_the_throwing_expression_equal_to_the_expected_throwable = \
5: @@ -186 +186 @@
5: -      is_the_result_of_the_throwing_expression_equal_to_the_expected_throwable = \
5: +    is_the_result_of_the_throwing_expression_equal_to_the_expected_throwable = \
5: 
5: 3 files with code style divergence
```

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15811/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10489/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13473/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16092/)
